src/main/java/com/google/firebase/database/connection/WebsocketConnection.java

```diff
@@ -16,43 +16,51 @@
 
 package com.google.firebase.database.connection;
 
-import com.google.firebase.database.connection.util.StringListReader;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.collect.ImmutableList;
 import com.google.firebase.database.logging.LogWrapper;
-import com.google.firebase.database.tubesock.WebSocket;
-import com.google.firebase.database.tubesock.WebSocketEventHandler;
-import com.google.firebase.database.tubesock.WebSocketException;
-import com.google.firebase.database.tubesock.WebSocketMessage;
 import com.google.firebase.database.util.JsonMapper;
 
 import java.io.EOFException;
 import java.io.IOException;
 import java.net.URI;
 import java.util.ArrayList;
-import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
-
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Represents a WebSocket connection to the Firebase Realtime Database. This abstraction acts as
+ * the mediator between low-level IO ({@link WSClient}), and high-level connection management
+ * ({@link Connection}). It handles frame buffering, most of the low-level errors, and notifies the
+ * higher layer when necessary. Higher layer signals this implementation when it needs to send a
+ * message out, or when a graceful connection tear down should be initiated.
+ */
 class WebsocketConnection {
 
   private static final long KEEP_ALIVE_TIMEOUT_MS = 45 * 1000; // 45 seconds
   private static final long CONNECT_TIMEOUT_MS = 30 * 1000; // 30 seconds
   private static final int MAX_FRAME_SIZE = 16384;
-  private static long connectionId = 0;
+  private static final AtomicLong CONN_ID = new AtomicLong(0);
```

Comment:
@mikelehen's comment regarding the suggested thread safety applies here as well.

Reply:
I almost called this out too, but actually since this is static it can technically be accessed by multiple threads (each Repo has its own run loop with its own thread and so these global connection ids could be assigned simultaneously on different threads).  So I think AtomicLong is applicable here (although a race is unlikely in practice and wouldn't be particularly harmful, since the ID is just used for logging).